presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+import com.facebook.presto.spi.type.ArrayType;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.spi.type.Type;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.util.Failures.checkCondition;
+
+@ScalarFunction("ngrams")
+@Description("Return N-grams for the input")
+public final class ArrayNgramsFunction
+{
+    private ArrayNgramsFunction() {}
+
+    @TypeParameter("T")
+    @SqlType("array(array(T))")
+    public static Block ngrams(@TypeParameter("T") Type type, @SqlType("array(T)") Block array, @SqlType(StandardTypes.BIGINT) long n)
```

Comment:
Consider putting the type parameter and array type construction in the constructor, like this:

```java
    private final Type type;
    private final ArrayType arrayType;

    @TypeParameter("T")
    public ArrayNgramsFunction(@TypeParameter("T") Type type) {
        this.type = type;
        this.arrayType = new ArrayType(type);
    }
```

And then this method becomes an instance method (remove static modifier).